text/2021-02-24-coprocessor-plugin.md

```diff
@@ -0,0 +1,241 @@
+# Coprocessor Plugin
+
+## Summary
+
+Add a general and pluggable coprocessor framework for RawKV mode.
+
+## Motivation
+
+TiKV is the storage component in the TiDB ecosystem, however, the distribution computation principle suggests that computation should be as close to the data source as possible. Therefore, TiKV has embedded a subset of the TiDB executor framework to push down some computation tasks when applicable.
+
+But TiKV's capability should be far beyond that, as many distributed components can be built on top of TiKV, such as cache, full text search engine, graph database and NoSQL database. And same as TiDB, these product will also like to push down specific computation to TiKV, which requires the coprocessor to be customizable, aka pluggable.
+
+For instance, a full-text seraching engine will persist the origin document and n-gram index on TiKV. It'll be a waste of resource if we read back and then update the index from a client. On the contrary, the coprocessor plugin can generate the index from the origin document, and update the index inplace. What's more, the coprocessor plugin can perform index scan directly on TiKV.
+
+The goals of the coprocessor plugin are:
+
+- Do what client can do
+
+- Easy to use
+  - Out of box
+  - Easy to deploy
+
+- Robust
+  - Easy to debug
+  - Log support, metrics support
+
+## Detailed design
```

Comment:
you can write the alternatives (not taken into implementation but meaningful for brainstorming and comparison) to the last sections, for example, the static mode for the plugin regrestering.

Reply:
Almost done, but I ignored `performance analysis, is there any performance penalty to use coprocessor plugin` because no benchmark no talk, and also ignored `milestones and sub-tasks` since it has been implemented.